src/main/java/io/github/opencubicchunks/cubicchunks/core/server/chunkio/region/ShadowPagingRegion.java

```diff
@@ -74,25 +81,79 @@ private ShadowPagingRegion(FileChannel file, SectorTracker<K> sectorTracker, IHe
 		this.sectorTracker = sectorTracker;
 	}
 
-	@Override public synchronized void writeValue(K key, ByteBuffer value) throws IOException {
+	/**
+	 * Writes an entry without updating the region headers on disk.
+	 * @return whether or not {@link #file} needs to be flushed before the headers are be updated
+	 */
+	protected boolean writeValueFirst(K key, ByteBuffer value) throws IOException {
 		if (value == null) {
 			this.sectorTracker.removeKey(key);
-			updateHeaders(key);
+			return false; //the data isn't actually changed on-disk, so we don't need to flush the channel before updating the headers
+		} else {
+			int size = value.remaining();
+			int sizeWithSizeInfo = size + Integer.BYTES;
+			int numSectors = this.getSectorNumber(sizeWithSizeInfo);
+			// this may throw UnsupportedDataException if data is too big
+			RegionEntryLocation location = this.sectorTracker.reserveForKey(key, numSectors);
+
+			int bytesOffset = location.getOffset() * this.sectorSize;
+
+			Utils.writeFully(this.file.position(bytesOffset), ByteBuffer.allocate(Integer.BYTES).putInt(0, size));
+			Utils.writeFully(this.file, value);
+			return true;
+		}
+	}
+
+	@Override public synchronized void writeValue(K key, ByteBuffer value) throws IOException {
+		if (this.writeValueFirst(key, value)) {
+			this.file.force(true);
+		}
+		this.updateHeaders(key);
+		//no need to flush channel a second time after headers update, as doing so won't affect data integrity
```

Comment:
Forst, your "else" in "writeValueFirst" makes the diff difficult to follow.

And I realized my old code was actually broken, and your code keeps it with a comment "explaining" that it's not broken. Not doing force() after the write for headers was an oversight on my part - I thought I didn't need to force() after a header write, because this tiny write could only either fail or succeed as a whole, and next write would force anyway. The issue is that I didn't think of the case where the next write could be replacing the data the previous header value was pointing to. And if that gets interrupted before force, before the new header data is written and the writes somehow get reordered (is that possible? I have no idea). you would get a header entry pointing to potentially incomplete write.

And honestly since this is internal code anyway - would any of your code break if you just made this write method throw an UnsupportedOperationException?

Reply:
ah, i didn't even consider that possibility. that'll definitely need a rework...

code that uses the non-batch `SaveSection#write` for whatever reason will still use this method, so making it throw `UnsupportedOperationException` isn't an option. what might make more sense is turning this into a proxy method (i.e. `writeValues(Collections.singletonMap(key, value))`)

Reply:
The point is that nothing in CC should be calling the non batch methods.